src/app/tabs/camera.cpp

```diff
@@ -1,135 +1,361 @@
-    layout->addStretch(1);
-
-    return widget;
-}
-
-void CameraTab::update_status(QMediaPlayer::MediaStatus media_status)
-{
-    qInfo() << "camera status changed to: " << media_status;
-
-    switch (media_status) {
-        case QMediaPlayer::LoadingMedia:
-        case QMediaPlayer::LoadedMedia:
-        case QMediaPlayer::BufferedMedia:
-            this->status->setText("connecting...");
-            break;
-        default:
-            this->status->setText("connection failed");
-            emit disconnected();
-            break;
-    }
-}
-
-void CameraTab::connect_stream()
-{
-    connect(this->player, &QMediaPlayer::mediaStatusChanged,
-            [this](QMediaPlayer::MediaStatus media_status) { this->update_status(media_status); });
-    connect(this->player, QOverload<>::of(&QMediaPlayer::metaDataChanged), [this]() { emit connected(); });
-
-    QString pipeline = QString(
-                           "gst-pipeline: rtspsrc location=%1 ! decodebin ! video/x-raw ! videoconvert ! videoscale ! "
-                           "xvimagesink sync=false force-aspect-ratio=false name=\"qtvideosink\"")
-                           .arg(this->url);
-
-    qInfo() << "playing stream pipeline: " << pipeline;
-    this->player->setMedia(QUrl(pipeline));
-    this->player->play();
-}
+#include <QLineEdit>
+#include <QCamera>
+#include <QCameraInfo>
+#include <QCameraImageCapture>
+
+#include <app/tabs/camera.hpp>
+#include <app/window.hpp>
+
+CameraTab::CameraTab(QWidget *parent) : QWidget(parent)
+{
+    this->theme = Theme::get_instance();
+    this->player = new QMediaPlayer(this);
+    this->local_cam = nullptr;
+    this->local_index = 0;
+
+    this->config = Config::get_instance();
+
+    QStackedLayout *layout = new QStackedLayout(this);
+    layout->setContentsMargins(0, 0, 0, 0);
+
+    layout->addWidget(this->connect_widget());
+    layout->addWidget(this->local_camera_widget());
+    layout->addWidget(this->network_camera_widget());
+
+    connect(this, &CameraTab::disconnected, [layout]() { layout->setCurrentIndex(0); });
+    connect(this, &CameraTab::connected_local, [layout]() { layout->setCurrentIndex(1); });
+    connect(this, &CameraTab::connected_network, [layout]() { layout->setCurrentIndex(2); });
+}
+
+QWidget *CameraTab::connect_widget()
+{
+    QWidget *widget = new QWidget(this);
+    QVBoxLayout *layout = new QVBoxLayout(widget);
+
+    QLabel *label = new QLabel("connect camera", widget);
+    label->setFont(Theme::font_16);
+
+    this->status = new QLabel(widget);
+    this->status->setFont(Theme::font_16);
+
+    QWidget *local_cams = this->camera_selector();
+    QWidget *network_cam = this->input_widget();
+    network_cam->setVisible(false);
+    QCheckBox *network_toggle = new QCheckBox("Network", this);
+    network_toggle->setFont(Theme::font_14);
+    connect(network_toggle, &QCheckBox::toggled, [this, network_cam, local_cams](bool checked) {
+        network_cam->setVisible(checked);
```

Comment:
something about this just feels wonky... like you're toggling between showing and hiding the two widgets?

Reply:
Yes, that's exactly the idea

Reply:
When network is checked I show the widget containing the network-specific options and when it's unchecked, the widget containing local options.

Reply:
it would probably be cleaner to have another nested stacked layout, and then just pick the layer that should shown at the time

Reply:
Ok, I changed it.